### PR TITLE
Feature: support endpoint name grouping by OpenAPI definitions.

### PR DESCRIPTION
**Pull request by wankai123**
### Support endpoint name grouping by OpenAPI definitions
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [X] Update the documentation to include this new feature.
- [X] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [X] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

Related to #7113
More information about the feature please see `docs/en/setup/backend/endpoint-grouping-rules.md`

### CODE REVIEW

File: oap-server/server-core/pom.xml
Comment (wankai123) [reply oap-server/server-core/pom.xml]: It's an omission, fixed

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:258]: I think these 3 scenarios are strange to describe.

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:287]: Move this as an example under the table should be good enough.

image

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:307]:
```suggestion

```

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:306]: This table is better than the above 3 cases.

Comment (kezhenxu94) [inline docs/en/setup/backend/endpoint-grouping-rules.md:306]: I'd rather put these GET:/products, etc. etc. into `` so you don't need to escape the characters, which is error prone

Comment (wu-sheng) [reply docs/en/setup/backend/endpoint-grouping-rules.md:306]: I am wondering the point of `\` too

Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md:306]: to escape the markdown keyword...  but if I put them into `` ,  the style (color) would differ from others.

Comment (wu-sheng) [reply docs/en/setup/backend/endpoint-grouping-rules.md:306]: Could you put a screenshot here to demonstrate what will be seen?

Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md:306]: image
would like this

Comment (wu-sheng) [reply docs/en/setup/backend/endpoint-grouping-rules.md:306]: @kezhenxu94 From the preview, it seems better to keep the current way?

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (kezhenxu94) [inline docs/en/setup/backend/endpoint-grouping-rules.md:280]: See
```suggestion
   | GET:/products/123 | serviceB | default | \${PATH}:\<\${METHOD}\> | true | /products/{id}:\<GET\> |
```

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (kezhenxu94) [inline docs/en/setup/backend/endpoint-grouping-rules.md:279]: See 
```suggestion
   | \<GET\>:/products/123 | serviceB | \<\${METHOD}\>:\${PATH} | \<\${METHOD}\>:\${PATH} | true | \<GET\>:/products/{id} |
```

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (kezhenxu94) [reply docs/en/setup/backend/endpoint-grouping-rules.md:281]: - `$`, `<`, and `>` has no special meanings in GitHub markdown, they may have in other extended markdown engine.

- You don't have consistent style in terms of quoting in backticks or not, for example you quote them here 
Screen Shot 2021-06-18 at 17 54 24
but not in this table

- If you choose not to quote in `` but with `\` to escape those characters, see 

- There are cases when users may look into the source codes and copy the example with those `\`, which doesn't work.

**Review by kezhenxu94 — DISMISSED** [review]
LGTM once CI pass. Thanks!!!

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md]: make sense, I have put all path/pattern string into ``.

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:28]: We should make this optional, and the default should be the folder name. The reason I am asking this is, once all things are not required, `x-sw-endpoint-name-match-rule`'s default value matched our ALS rule(@kezhenxu94 it is matched, right?), we could only ask for standard OpenAPI spec file in the right folder, nothing more.

Comment (kezhenxu94) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: Yes they matched

Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: make sense

Comment (wu-sheng) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: > But if the folder is nested,

What do you mean `folder nested`? I think you should require a folder as the service name, but could use `x-sw-service-name` to override.

Comment (kezhenxu94) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: The folder can be any number depth


```
a
 |--b
 |--|--c
 |--|--|--d
```

Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: A recommend practice from OAS is `A good rule of thumb is to use the natural hierarchy present in URLs to build your file structure. For example, put all routes starting with /users (like /users and /users/{id}) in the same file (think of it as a “sub-API”).`

Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: I think using the folder name that the file belong is enough.

File: oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/ResourceUtils.java
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/ResourceUtils.java:73]: Notice, this kind of `Recursive` could easily cause OOM. Like I commented https://github.com/apache/skywalking/pull/7130#discussion_r654422789, please make the structure solid.

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: Read https://github.com/apache/skywalking/pull/7130#discussion_r654422789

File: oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/ResourceUtils.java
Comment (wu-sheng) [reply oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/ResourceUtils.java:73]: We should have depth as a parameter, as in this OpenAPI case, we only need recursive depth as `1`, which means current folder and one more(service name folder).

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md]: done

File: oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/ResourceUtils.java
Comment (wankai123) [reply oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/ResourceUtils.java]: done

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:38]: Let's keep these as same as default value, to avoid any confusion. 

@wankai123 We also need an issue and another PR to fix current operation name in plugins. You should fix those after this gets merged.

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:59]: As no side effect, keeping this open in default should be better?

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:28]: What happens in no `x-sw-service-name` and `yaml` file is located in the `openapi-definitions` directory directly.

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:50]:
```suggestion
1. All OpenAPI definition documents should be located in the `openapi-definitions` directory, with at most two levels directories. Recommend using the service name as the subDirectory name then you are not necessary to set `x-sw-service-name`. For example:
```

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wankai123) [reply docs/en/setup/backend/endpoint-grouping-rules.md:28]: `openapi-definitions` would be the service name, should throw an exception?

File: docs/en/setup/backend/endpoint-grouping-rules.md
Comment (wu-sheng) [inline docs/en/setup/backend/endpoint-grouping-rules.md:17]:
```suggestion
The OpenAPI definitions are the documents based on The [OpenAPI Specification (OAS)](https://www.openapis.org/) which used to define a standard, language-agnostic interface for HTTP APIs.
```

Once we say a spec, we should link to their website rather that GitHub repo. GitHub is for developers.

**Review by wu-sheng — DISMISSED** [review]
It is good for me. Thanks.

File: oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java
Comment (kezhenxu94) [inline oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java:51]: Not important though, `"GET:/products1/{id}/" + +i` looks tricky, why not just `GET:/products1/{id}/" + i`?

File: oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java
Comment (kezhenxu94) [inline oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java:72]: Same here

File: oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java
Comment (kezhenxu94) [inline oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java:93]: And here

**Review by kezhenxu94 — DISMISSED** [review]

File: oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java
Comment (wu-sheng) [reply oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java:51]: Haha, interesting, it is better to fix before merge.

Comment (wankai123) [reply oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/config/group/openapi/EndpointGroupingBenchmark4Openapi.java]: thanks, @kezhenxu94, I tested `+i` but forgot to remove them.

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**
